dr: add the `odf dr validate clusters` command (#185)
red-hat-storage/odf-cli: docs/dr.md

validate application command. If the command failed, check the error details in
the log.

> ### validate clusters

Reviewer: ```suggestion
### Validate clusters
```

Author: This is not consistent with other commands sections in this document:
- https://github.com/red-hat-storage/odf-cli/blob/main/docs/dr.md#test-run
- https://github.com/red-hat-storage/odf-cli/blob/main/docs/dr.md#test-clean
- https://github.com/red-hat-storage/odf-cli/blob/main/docs/dr.md#validate-application
- https://github.com/red-hat-storage/odf-cli/blob/main/docs/dr.md#gather-application

The validate clusters command validates the disaster recovery clusters by
gathering cluster scoped and related ramen resources from all clusters.

### Validating clusters

To validate the disaster recovery clusters, run the following command:

```console
$ odf dr validate clusters -o out
⭐ Using config "config.yaml"
⭐ Using report "out"

🔎 Validate config ...
   ✅ Config validated

🔎 Validate clusters ...
   ✅ Gathered data from cluster "hub"
   ✅ Gathered data from cluster "dr1"
   ✅ Gathered data from cluster "dr2"
   ✅ Clusters validated

✅ Validation completed (36 ok, 0 stale, 0 problem)
```

The command gathered cluster scoped and ramen resources from all clusters,
inspected the resources, and stored output files in the specified output
directory:

```console
$ tree -L1 out
out
├── validate-clusters.data
├── validate-clusters.log
└── validate-clusters.yaml
```

> [!IMPORTANT]
> When reporting DR related issues, please create an archive with the output
> directory and upload it to the issue tracker.

> #### The validate-clusters.yaml

Reviewer: ```suggestion
#### Validate clusters file
```
?

Author: This is not consistent with other comamnds:
- https://github.com/red-hat-storage/odf-cli/blob/main/docs/dr.md#the-test-runyaml
- https://github.com/red-hat-storage/odf-cli/blob/main/docs/dr.md#the-validate-applicationyaml

We can change all title to:

```
#### The `validate-clusters.yaml` file
```

Which renders like:

#### The `validate-clusters.yaml` file

But I need to get the command in the 4.20 branch first and backport to developer preview change to 4.19 first.

Lets consider this for next upstream work when the 4.20 release is done.

The `validate-clusters.yaml` report is a machine and human readable description
of the command and the clusters status.

> The most important part of the report is the clustersStatus:

Reviewer: ```suggestion
The most important part of the report is the clusters status:
```
?

Author: Pushing an update for this and applicationStatus...

    drPolicies:
      state: ok ✅
      value:
      - conditions:
        - state: ok ✅
          type: Validated
        drClusters:
        - dr1
        - dr2
        name: dr-policy
        schedulingInterval: 1m
    ramen:
      configmap:
        deleted:
          state: ok ✅
        name: ramen-hub-operator-config
        namespace: ramen-system
        ramenControllerType:
          state: ok ✅
          value: dr-hub
        s3StoreProfiles:
          state: ok ✅
          value:
          - s3ProfileName: minio-on-dr1
            s3SecretRef:
              state: ok ✅
              value:
                name: ramen-s3-secret-dr1
                namespace: ramen-system
          - s3ProfileName: minio-on-dr2
            s3SecretRef:
              state: ok ✅
              value:
                name: ramen-s3-secret-dr2
                namespace: ramen-system
      deployment:
        conditions:
        - state: ok ✅
          type: Available
        - state: ok ✅
          type: Progressing
        deleted:
          state: ok ✅
        name: ramen-hub-operator
        namespace: ramen-system
        replicas:
          state: ok ✅
          value: 1
```

> #### The validate-clusters.data directory

Reviewer: ```suggestion
####  Validate clusters data directory
```

│   │   ├── cluster.open-cluster-management.io
│   │   ├── flowcontrol.apiserver.k8s.io
│   │   ├── namespaces
│   │   ├── networking.k8s.io
│   │   ├── nodes
│   │   ├── operator.open-cluster-management.io
│   │   ├── operators.coreos.com
│   │   ├── persistentvolumes
│   │   ├── ramendr.openshift.io
│   │   ├── rbac.authorization.k8s.io
│   │   ├── replication.storage.openshift.io
│   │   ├── scheduling.k8s.io
│   │   ├── snapshot.storage.k8s.io
│   │   ├── storage.k8s.io
│   │   ├── submariner.io
│   │   └── work.open-cluster-management.io
│   └── namespaces
│       └── ramen-system
└── hub
    ├── cluster
    │   ├── addon.open-cluster-management.io
    │   ├── admissionregistration.k8s.io
    │   ├── apiextensions.k8s.io
    │   ├── apiregistration.k8s.io
    │   ├── cluster.open-cluster-management.io
    │   ├── flowcontrol.apiserver.k8s.io
    │   ├── namespaces
    │   ├── networking.k8s.io
    │   ├── nodes
    │   ├── operator.open-cluster-management.io
    │   ├── operators.coreos.com
    │   ├── ramendr.openshift.io
    │   ├── rbac.authorization.k8s.io
    │   ├── scheduling.k8s.io
    │   └── storage.k8s.io
    └── namespaces
        └── ramen-system
```

> #### The validate-clusters.log

Reviewer: ```suggestion
#### Validate clusters log
```

Author: Same, we want to mention the exact name.

Reviewer: I'm not sure if I have seen heading with above format that include file name.

Author: This document already include such headings, this is also in upstream docs:
https://github.com/RamenDR/ramenctl/blob/main/docs/validate.md

We want to keep the odf dr same as upstream docs to make it easier to keep them synced. The difference in the command names (`odf dr` instead of `ramenctl`) and sometimes in the example applications, we show results from OCP clusters if possible.

This log includes detailed information that may help to troubleshoot the
validate clusters command. If the command failed, check the error details in

> the log.

Reviewer: ```suggestion
the log.

```

Author: Do you want to add an empty line at the end of the document? why?

Author: Ok, pushing a fix, thanks for noticing!
